### DISCUSSION TITLE
How to search through the history of a file

### DISCUSSION THREAD

**Discussion by pvonmoradi** [jesseduffield/lazygit#1377] [Q&A]
Is it possible to search through all lines of the history of a *file* (not commits)? For example to find out *when* a particular string was added to a file then `checkout` to that commit?

**Comment by jordan-yee** [comment]
I'm not sure if this is quite what you're after, but it's what I was looking for when I clicked on this question:

**`<ctrl-s>` allows you to enter a path to a file to filter on.**

Once you've entered a path, you can then navigate through a list of commits in which that file was changed, with the changes to that file being displayed on the right-hand side. If you use `<ctrl-s>` when a file is selected in the Files region, then you have the option to filter on that file right away.

My problem was that I wanted to check a file that wasn't shown in the changed files list, and I wasn't quite sure what the exact path was to type into the prompt. This turns out to be a good place where the `:` key (to enter an arbitrary shell command) is useful. This allows you to run any external utility you want to accomplish the task. I use the ranger application (installed separately) to browse files, so I could just open that from lazygit by typing `:ranger` in lazygit. This opens ranger, from which I can quickly navigate the file I want, copy the path with a keystroke, then exit. After exiting an external program, you're returned to the command line with a "press enter to return to lazygit" message. From here I could press enter, use the paste function of my terminal emulator (the text input prompt in lazygit doesn't seem to offer any editing keybindings), then proceed to view the change history of the target file in lazygit.

An even better solution would be a command to use fzf (a fuzzy search tool) to fuzzy search the directory for your file and copy the path to your clipboard. This is getting off-topic, but maybe it will be useful for someone trying to optimize their workflow surrounding this feature...

**Reply by pvonmoradi** [reply]
Thanks for this post. I was not aware of `ctrl+s` and indeed it shows history of changes for a file.  
1- Still, I can't search through the contents of the file or the diffs (patches) in right pane.
2- For a workaround, we can use `git log -p path/to/file` which pipes the diffs to `less` so the added/deleted lines are searchable (with `/`). This way, we can use fzf (`ctrl+t` in shell) to locate the file.

**Reply by screendriver** [reply]
Same here: thank you for that post @jordan-yee! This was exactly I was looking for 👍 My workflow is now almost the same as already explained here except that I'm using [nnn](https://github.com/jarun/nnn) as file manager to fuzzy find and copy the path to the file for what I want so see the history. It's a little bit more effort (keystrokes) as it would be when fuzzy finding files would already be included in `lazygit` directly but it works as expected 👍 

@jesseduffield are there any plans to include a fuzzy finder like [fzf](https://github.com/junegunn/fzf) directly in `lazygit`?

**Reply by jesseduffield** [reply]
@screendriver it's a feature that would be good to have. I don't believe we have an issue for this yet: would you be able to create an issue describing the desired behaviour?

**Reply by screendriver** [reply]
Done: https://github.com/jesseduffield/lazygit/issues/1525

**Comment by prodehghan** [comment]
@pvonmoradi The git's "pickaxe" command comes handy for this. Look at [here](https://git-scm.com/book/en/v2/Git-Tools-Searching#:~:text=currently%20checked%20out.-,Git%20Log%20Searching,-Perhaps%20you%E2%80%99re%20looking).
